### PR TITLE
Support for multiple OAuth providers

### PR DESCRIPTION
**Pull request by hhorne**
Support for allowing the current user to link and unlink multiple OAuth providers from their account.

### CODE REVIEW

File: src/auth.test.js [hunk 1]
```diff
@@ -299,6 +302,39 @@ describe('auth', () => {
...
+
+      expect(iterator.next().value)
+        .toEqual(call([auth.currentUser, auth.currentUser.linkWithPopup], authProvider))
+
+      expect(iterator.next({ credential })).toEqual({
+        done: true,
+        value: { credential }
+      })
+    })
+  })
```
Comment (n6g7) [inline src/auth.test.js]: We should also add a test suite for `linkWithRedirect`. ;)

File: docs/_reference/auth.md [hunk 2]
```diff
@@ -98,6 +98,53 @@ methods:
       }
       ```
 
+  - signature: auth.linkWithPopup(authProvider)
+    id: linkWithPopup
+    generator: true
+    description: Links the authenticated provider to the user account using a pop-up based OAuth flow.
+    arguments:
+      - name: authProvider
+      required: true
```
Comment (n6g7) [inline docs/_reference/auth.md]: Indentation is broken here

File: docs/_reference/auth.md [hunk 3]
```diff
@@ -98,6 +98,53 @@ methods:
...
+      }
+      ```
+
+  - signature: auth.linkWithRedirect(authProvider)
+    id: linkWithRedirect
+    generator: true
+    description: Links the authenticated provider to the user account using a full-page redirect flow.
+    arguments:
+      - name: authProvider
+      required: true
```
Comment (n6g7) [inline docs/_reference/auth.md]: indentation

File: docs/_reference/auth.md [hunk 4]
```diff
@@ -98,6 +98,53 @@ methods:
...
+      type: A [firebase.auth.AuthProvider](https://firebase.google.com/docs/reference/js/firebase.auth.AuthProvider) object.
+      description: The authentication provider to use for the request.
+    output: A [firebase.Promise](https://firebase.google.com/docs/reference/js/firebase.Promise) containing void.
```
Comment (n6g7) [inline docs/_reference/auth.md]: The promise will be resolved by redux-saga so the saga will directly return `null`/`undefined`. You can remove this line. 😉

File: docs/_reference/auth.md [hunk 5]
```diff
@@ -98,6 +98,53 @@ methods:
...
+      type: A [firebase.auth.AuthProvider](https://firebase.google.com/docs/reference/js/firebase.auth.AuthProvider) object.
+      description: The authentication provider to use for the request.
+    output: A [firebase.auth.UserCredential](https://firebase.google.com/docs/reference/js/firebase.auth#.UserCredential) instance.
+    example: |
+      ```javascript
+      const authProvider = new firebase.auth.GoogleAuthProvider();
+
+      function* linkSaga() {
+        try {
+          const data = yield(rsf.auth.linkWithPopup, authProvider);
```
Comment (n6g7) [inline docs/_reference/auth.md]: You forgot the `call` function. 🙂

File: docs/_reference/auth.md [hunk 6]
```diff
@@ -98,6 +98,53 @@ methods:
...
+      type: A [firebase.auth.AuthProvider](https://firebase.google.com/docs/reference/js/firebase.auth.AuthProvider) object.
+      description: The authentication provider to use for the request.
+    output: A [firebase.Promise](https://firebase.google.com/docs/reference/js/firebase.Promise) containing void.
+    example: |
+      ```javascript
+      const authProvider = new firebase.auth.GoogleAuthProvider();
+
+      function* linkSaga() {
+        try {
+          yield(rsf.auth.linkWithRedirect, authProvider);
```
Comment (n6g7) [inline docs/_reference/auth.md]: `call` 😉

**Comment by n6g7** [comment]
Looks like you'll also have to rebase on master to trigger the circleci builds with the up-to-date config. ;)

File: src/auth.test.js (hunk 1 above)
Comment (hhorne) [reply src/auth.test.js]: All done!

**Comment by hhorne** [comment]
I think that's everything except the rebase. I tried rebasing but it didn't seem to do what was needed. I'm sure it's a mistake on my part.

**Comment by hhorne** [comment]
All checks passed now but based on the list of commit above I feel like i may have messed something up...

**Comment by hhorne** [comment]
Closing this PR and opening another with less screwed up history.

**Pull request closed without merging**